### ISSUE TITLE
Multi level caching suggestion?

### ISSUE THREAD

**Issue by aphillipo** [cabol/nebulex#168]
Hi loving Nebulex it's really fantastic!

```elixir
@decorate cacheable(cache: Cache, key: {User, sub}, mapped_to: fn user -> {User, user.id} end, opts: [ttl: @ttl])
def get_user_by_sub!(sub),
    do: Repo.get_by!(User, %{sub: sub})
```

This would mean instead of adding an entry to the cache for that specific {User, sub} it became multi layered and it would store in the main cache {User, id}, updating it there should this be needed and also adding a reference to this value from {User, sub}. This means that we can now just delete {User, id} from the cache with cache_evict and not have to track each type of lookup.

Taking the following as a further example of how mapped_to could be useful we could do the following:

```elixir
@decorate cacheable(
    cache: Cache, 
    key: {User, ids}, # something clever here maybe?
    mapped_to: fn users -> 
        for u <- users, do: {User, u.id} # mapped to can be an array entry for each item in ids.
    end, 
    opts: [ttl: @ttl]
)
def get_users_by_ids(ids),
    do: from(users in User, where: users.id in ^ids) |> Repo.all()
```

And I think there is (nearly) a enough info for the above to hit the database only for the values not in the mapped cache list already?

Probably this is a bit complex and I shouldn't worry, just would have been nice with what I'm building to know I wasn't hitting the DB rather than the cache to update the leaderboard as participants enter and leave a classroom. Maybe there is already a way to do this in Nebulex I would not be surprised!

WDYT?

**Comment by cabol** [comment]
Hey 👋 !! That's actually a very good and interesting suggestion, it definitely would be very useful, I think it really worth exploring it. Let me try to come up with a design for this, but also feel free to open a PR or put your thoughts/ideas here. Thanks for the suggestion, it is pretty good! Will try to give you feedback soon!

**Comment by aphillipo** [comment]
Yes if I get a chance next week I'll have a look at the source code and have a play...

**Comment by cabol** [comment]
I've created a separate issue describing the feature request, please check it out and let me know if I miss something (#169). BTW, will open a new branch with the solution soon so you can try it out. Thanks!

**Comment by cabol** [comment]
Closing this issue since I haven't received feedback for a long time ago. Feel free to reopen it or create a new one if the implemented feature (#169) doesn't solve it.
